finetune/model.py

```diff
@@ -136,14 +149,39 @@ def _finetune(self, *Xs, Y, batch_size=BATCH_SIZE):
         self._build_model(n_updates_total=n_updates_total, n_classes=self.n_classes)
 
         dataset = shuffle(train_x, train_mask, Y, random_state=np.random)
+        x_tr, x_va, m_tr, m_va, y_tr, y_va = train_test_split(*dataset, test_size=val_frac, random_state=31415)
+
+        dataset = (x_tr, m_tr, y_tr)
+        val_dataset = (x_va, m_va, y_va)
 
         self.is_trained = True
-        rolling_avg_loss = 0
+        avg_train_loss = 0
+        avg_val_loss = 0
+        global_step = 0
+        best_val_loss = float("inf")
         for i in range(N_EPOCHS):
             for xmb, mmb, ymb in iter_data(*dataset, n_batch=n_batch_train, verbose=True):
-                cost, _ = self.sess.run([self.clf_loss, self.train_op], {self.X: xmb, self.M: mmb, self.Y: ymb})
-                rolling_avg_loss = rolling_avg_loss * 0.95 + cost *  0.05
-                print("\nLOSS = {}, ROLLING AVG = {}".format(cost, rolling_avg_loss))
+                global_step += 1
+                if global_step % eval_interval == 0:
+                    sum_val_loss = 0
+                    for xval, mval, yval in enumerate(iter_data(*val_dataset, n_batch=n_batch_train, verbose=True)):
+                        val_cost, summary = self.sess.run([self.clf_loss, self.summaries],
+                                                          {self.X: xval, self.M: mval, self.Y: yval})
+                        self.valid_writer.add_summary(summary, global_step)
+                        sum_val_loss += val_cost
+                        avg_val_loss = avg_val_loss * ROLLING_AVG_DECAY + val_cost * (1 - ROLLING_AVG_DECAY)
+                        print("\nVAL: LOSS = {}, ROLLING AVG = {}".format(val_cost, avg_val_loss))
```

Comment:
Would recommend either using the verbose parameter to decide whether to write to stdout or use the logging module with a logging.INFO level so folks can opt out of getting stdout output.